siteapp/views.py

```diff
@@ -1393,31 +1393,86 @@ def portfolio_list(request):
-        logger.info(
-            event="new_portfolio assign_owner_permissions",
-            object={"object": "portfolio", "id": portfolio.id, "title":portfolio.title},
-            receiving_user={"id": request.user.id, "username": request.user.username},
-            user={"id": request.user.id, "username": request.user.username}
-        )
-        return redirect('portfolio_projects', pk=portfolio.pk)
+        try:
+            form = PortfolioForm(request.POST)
+            if form.is_valid():
+                form.save()
+                portfolio = form.instance
+                logger.info(
+                    event="new_portfolio",
+                    object={"object": "portfolio", "id": portfolio.id, "title": portfolio.title},
+                    user={"id": request.user.id, "username": request.user.username}
+                )
+                portfolio.assign_owner_permissions(request.user)
+                logger.info(
+                    event="new_portfolio assign_owner_permissions",
+                    object={"object": "portfolio", "id": portfolio.id, "title": portfolio.title},
+                    receiving_user={"id": request.user.id, "username": request.user.username},
+                    user={"id": request.user.id, "username": request.user.username}
+                )
+                return redirect('portfolio_projects', pk=portfolio.pk)
+        except IntegrityError:
+            messages.add_message(request, messages.ERROR, "Portfolio name not available." )
     else:
         form = PortfolioForm()
-
     return render(request, 'portfolios/form.html', {
         'form': form,
         "project_form": ProjectForm(request.user),
     })
 
+@login_required
+def delete_portfolio(request, pk):
+    """Form to delete portfolios"""
+    if request.method == 'GET':
+        portfolio = Portfolio.objects.get(pk=pk)
+        logger.info(
+            event="delete_portfolio",
+            object={"object": "portfolio", "id": portfolio.id, "title": portfolio.title},
+            user={"id": request.user.id, "username": request.user.username}
+        )
+        Portfolio.objects.get(pk=pk).delete()
+        return redirect("list_portfolios")
+
+
+@login_required
+def edit_portfolio(request, pk):
+    """Form to edit portfolios"""
+
+    if request.method == 'GET':
+        portfolio = Portfolio.objects.get(pk=pk)
+        form = PortfolioForm(request.POST or None, instance=portfolio, initial={'portfolio': portfolio.id})
+        if form.is_valid():
+            form.save()
+
+            logger.info(
+                event="edit_portfolio",
+                object={"object": "portfolio", "id": portfolio.id, "title": portfolio.title},
+                user={"id": request.user.id, "username": request.user.username}
+            )
+            portfolio.assign_owner_permissions(request.user)
+            logger.info(
+                event="new_portfolio assign_owner_permissions",
```

Comment:
Is "new_portfolio" needed here? Since it's just editing the portfolio, should that part be removed to just say "assign_owner_permissions"? Especially since an "edit_portfolio" event is logged just 2 lines before.

Reply:
Yup permissions can be assigned through other functions